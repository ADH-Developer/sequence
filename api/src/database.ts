import { Sequelize, SequelizeOptions } from "sequelize-typescript";

const sequelizeOpts: SequelizeOptions = {
  dialectOptions: {
    ssl: {
      require: true,
      rejectUnauthorized: false,
    },
  },
  logging: process.env.NODE_ENV === "local" ? true : false,
};

const sequelize = new Sequelize(
  process.env.DATABASE_URL,
  process.env.NODE_ENV === "development" ||
  process.env.NODE_ENV === "production"
    ? sequelizeOpts
    : {}
);

export default sequelize;
