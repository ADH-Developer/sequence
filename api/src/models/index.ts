import User from "./user";
import AuthGoogle from "./auth_google";
import sequelize from "../database";

async function buildModels() {
  try {
    console.log("syncing models");
    await sequelize.sync({ force: true });
    // await sequelize.sync({});
  } catch (error) {
    console.error(error);
  }
}
export default buildModels;
