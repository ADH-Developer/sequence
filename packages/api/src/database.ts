import { Sequelize, SequelizeOptions } from "sequelize-typescript";
import SequelizeConfig from "./config/config";

declare type ENVIRONMENT = "test" | "development" | "production";

const options: SequelizeOptions = {
  dialectOptions: process.env.DB_SSL === "true"
    ? {
      ssl: {
        require: true,
        rejectUnauthorized: false,
      },
    }
    : undefined,
  logging: process.env.DB_LOGGING === "true" ? console.log : false,
};

const config = {
  url: process.env.DB_URL,
  username: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
  host: process.env.DB_HOST,
  port: parseInt(process.env.DB_PORT || "5432"),
  dialect: "postgres",
  ...options
};

export const dbConfig = config;

const createSequelize = (...args: any[]) => {
  const instance = new Sequelize(...args);
  instance
    .authenticate()
    .then(() => console.log("Sequence: database connection successful"))
    .catch((error) => {
      console.error("Failed to connect to database:", error.message);
    });
  return instance;
};

let sequelize: Sequelize;
if (config.url) {
  sequelize = createSequelize(config.url, { ...config, ...options });
} else {
  sequelize = createSequelize({ ...config, ...options });
}

export default sequelize;
