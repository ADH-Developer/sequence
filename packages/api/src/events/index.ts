export * from "./identify.event";
export * from "./track.event";

export type SegmentEventTypes =
  | "track"
  | "identify"
  | "alias"
  | "page"
  | "group";

export type SegmentBase = {
  type: SegmentEventTypes;
  anonymousId: string;
  context?: SegmentContext;
  userId?: string;
  receivedAt: string;
  timestamp: string;
  sentAt: string;
  messageId: string;
  integrations?: Record<string, boolean>;
};

interface SegmentContext {
  library?: {
    sdk: string;
    name: string;
    version: string;
  };
  page?: {
    path: string;
    referrer: string;
    search: string;
    title: string;
    url: string;
  };
  userAgent?: string;
  ip?: string;
}

export interface SegmentTrack extends SegmentBase {
  type: "track";
  event: string;
  originalTimestamp?: string;
  properties: Record<string, any>;
}

export interface SegmentIdentify extends SegmentBase {
  type: "identify";
  channel?: string;
  traits: Record<string, any>;
}

export interface SegmentPage extends SegmentBase {
  type: "page";
  name: string;
  properties: {
    title: string;
    url: string;
  };
}
