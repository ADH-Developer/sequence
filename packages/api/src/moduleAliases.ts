const alias = require("module-alias");

export const register = () => {
  if (process.env.BUILD || process.env.NODE_ENV === "production") {
    alias.addAliases({
      src: "./src/",
      tests: "./tests/",
    });
  } else {
    alias.addAliases({
      src: "./src/",
      tests: "./tests/",
    });
  }
};
