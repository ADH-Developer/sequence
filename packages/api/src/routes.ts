import { Application } from "express";
import passport from "passport";
import { authRoutes } from "./routes/auth.routes";
import { webhookRoutes } from "./routes/webhook.routes";
import { testRoutes } from "./routes/test.routes";

class Routes {
    constructor(app: Application) {
        app.use(passport.initialize());
        app.use("/auth", authRoutes);
        app.use("/webhooks", webhookRoutes);
        app.use("/test", testRoutes);
    }
}

export default Routes; 