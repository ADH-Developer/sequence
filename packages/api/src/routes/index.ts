import { Application } from "express";
import Passport from "./passport.http";
import SegmentHttpHandler from "./segment.http";
import SequenceHttpHandler from "./sequence.http";
import MiddlewareRegistry from "../services/middleware.registry";

class HttpRoutes {
  constructor(app: Application) {
    MiddlewareRegistry.addMiddleware(new Passport(app));
    MiddlewareRegistry.addMiddleware(new SegmentHttpHandler(app));
    MiddlewareRegistry.addMiddleware(new SequenceHttpHandler(app));
  }
}

export default HttpRoutes;
