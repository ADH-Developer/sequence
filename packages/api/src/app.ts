import dotenv from "dotenv";
dotenv.config();
import { register } from "./moduleAliases";
register();
import express from "express";
import cookieParser from "cookie-parser";
import { ApolloServer } from "apollo-server-express";
import jwt from "jsonwebtoken";
import cors from "cors";
import session from "express-session";
import enforce from "express-sslify";
import JwtConfig from "./config/jwt";
import schema from "./graphql/schema";
import SequelizeModels from "./models/index";
import Routes from "./routes";
import resolvers from "./graphql/resolvers";
import logger from "./utils/logger";

const app = express();

app.disable("x-powered-by");

if (process.env.USE_SSL) {
  app.use(enforce.HTTPS({ trustProtoHeader: true }));
}

app.use(
  cors({
    origin: [
      "https://my.sequence.so",
      "https://dev.sequence.so",
      "http://localhost:8000",
    ],
    credentials: true,
  })
);

app.options("*", cors());

app.use(cookieParser(process.env.JWT_SECRET_KEY));
app.use(express.json());
app.use(
  express.urlencoded({
    extended: true,
  })
);
app.use(
  session({
    secret: process.env.ENCRYPTION_KEY,
    resave: true,
    saveUninitialized: true,
  })
);

new Routes(app);

const server = new ApolloServer({
  typeDefs: schema,
  resolvers: resolvers,
  context: ({ req }) => {
    const token = req.headers.authorization;
    const result = jwt.verify(token, JwtConfig.jwt.secret);
    const user = (result as any).user;
    return {
      models: SequelizeModels,
      user,
    };
  },
  formatError(err) {
    logger.error("Error occured processing request:", err);
    return {
      message: err.message,
      // code: err.originalError && err.originalError.code,
      locations: err.locations,
      path: err.path,
    };
  },
});

server.applyMiddleware({
  app,
  cors: {
    origin: [
      "https://my.sequence.so",
      "https://dev.sequence.so",
      "http://localhost:8000",
    ],
    credentials: true,
  },
});

export default app;
