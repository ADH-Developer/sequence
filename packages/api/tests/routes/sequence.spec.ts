import { expect } from "chai";
import fetch from "node-fetch";
import App from "src/app";
import { SegmentIdentify, SegmentTrack } from "src/events";
import Event from "src/models/event.model";
import ProductUser from "src/models/productUser.model";
import SegmentWebhook from "src/models/segmentWebhook.model";
import User from "src/models/user.model";
import sequenceWebhookSeed from "tests/seeds/sequenceWebhook.seed";
import userSeed from "tests/seeds/user.seed";
import {
  courseClicked,
  identify1,
  ecommerceOrderCompleted,
  trackExperimentViewed,
  identify3,
  identify2,
} from "./eventData";

let user: User;
let webhook: SegmentWebhook;
let app: App;

describe("sequence webhook", () => {
  before(async () => {
    user = await userSeed();
    webhook = await sequenceWebhookSeed({
      userId: user.id,
    });
    app = new App({
      port: 3000,
    });
    await app.listen();
  });
  after(async () => {
    await Event.destroy({
      where: {
        userId: user.id,
      },
      force: true,
    });
    await ProductUser.destroy({
      where: {
        userId: user.id,
      },
      force: true,
    });
    await user.destroy();
    await webhook.destroy();
    app.getServer().close();
  });

  const performAPI = async (
    data: any,
    url = "http://localhost:3000/event/batch"
  ) => {
    // 1. Get Token from onboarding page
    const basicUsername = webhook.token;

    // 2. Base64 encode the Username:Password (password is empty)
    const token = Buffer.from(basicUsername + ":").toString("base64");

    // 3. Add your token to `Authorization` in the header
    const res = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Basic ${token}`,
      },
      body: JSON.stringify(data),
    });
    const json = await res.json();
    return json;
  };

  describe("sequence import", () => {
    it("Course Clicked- Intro to Analytics", async () => {
      const payload = { ...courseClicked, sentAt: new Date().toUTCString() };
      const result = await performAPI({
        batch: [payload as SegmentTrack],
      });
      expect(result).toMatchSnapshot();
      const event = await Event.findOne({
        where: {
          messageId: courseClicked.messageId,
          userId: user.id,
        },
      });
      expect(payload.context).to.deep.eq(event.context);
      expect(payload.properties).to.deep.eq(event.properties);
    });
  });
  describe("identify", () => {
    it("Identify 1", async () => {
      const payload = { ...identify1, sentAt: new Date().toUTCString() };
      const result = await performAPI({
        batch: [payload as SegmentIdentify],
      });
      expect(result).toMatchSnapshot();
      const person = await ProductUser.findOne({
        where: {
          externalId: payload.userId,
          userId: user.id,
        },
      });
      expect(payload.traits.firstName).to.deep.eq(person.firstName);
      expect(payload.traits.lastName).to.deep.eq(person.lastName);
    });
  });

  describe("mixed", () => {
    it("batch import many with idempotency", async () => {
      const result1 = await performAPI({
        batch: [
          ecommerceOrderCompleted,
          trackExperimentViewed,
          identify2,
          identify3,
        ],
      });
      expect(result1).toMatchSnapshot();
      const result2 = await performAPI({
        batch: [
          ecommerceOrderCompleted,
          trackExperimentViewed,
          identify2,
          identify3,
        ],
      });
      expect(result2).toMatchSnapshot();
    });
  });

  it("should create a user", async () => {
    const userData = {
      firstName: "Wei",
      lastName: "James",
      email: "pifagor@example.me",
      plan: "premium",
      logins: 5,
      myCustomAttribute: "testing",
      externalId: "kfzKyjdURh",
    };
    const result = await performAPI(userData, "http://localhost:3000/user");
    delete result.batch[0].messageId;
    expect(result).toMatchSnapshot();
  });
});
