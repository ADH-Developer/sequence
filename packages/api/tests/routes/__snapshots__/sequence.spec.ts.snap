// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sequence webhook identify Identify 1 1`] = `
Object {
  "batch": Array [
    Object {
      "messageId": "AcxAMBYg5x",
      "processed": true,
      "success": true,
    },
  ],
  "errors": 0,
  "processed": 1,
  "success": true,
  "total": 1,
}
`;

exports[`sequence webhook mixed batch import many with idempotency 1`] = `
Object {
  "batch": Array [
    Object {
      "messageId": "z9WKlllaNM",
      "processed": true,
      "success": true,
    },
    Object {
      "messageId": "rQeUO6L9yV",
      "processed": true,
      "success": true,
    },
    Object {
      "messageId": "kfzKyjdURh",
      "processed": true,
      "success": true,
    },
    Object {
      "messageId": "ILuRjMsgnJ",
      "processed": true,
      "success": true,
    },
  ],
  "errors": 0,
  "processed": 4,
  "success": true,
  "total": 4,
}
`;

exports[`sequence webhook mixed batch import many with idempotency 2`] = `
Object {
  "batch": Array [
    Object {
      "messageId": "z9WKlllaNM",
      "processed": false,
      "success": true,
    },
    Object {
      "messageId": "rQeUO6L9yV",
      "processed": false,
      "success": true,
    },
    Object {
      "messageId": "kfzKyjdURh",
      "processed": true,
      "success": true,
    },
    Object {
      "messageId": "ILuRjMsgnJ",
      "processed": true,
      "success": true,
    },
  ],
  "errors": 0,
  "processed": 2,
  "success": true,
  "total": 4,
}
`;

exports[`sequence webhook sequence import Course Clicked- Intro to Analytics 1`] = `
Object {
  "batch": Array [
    Object {
      "messageId": "DjqvhmyoSH",
      "processed": true,
      "success": true,
    },
  ],
  "errors": 0,
  "processed": 1,
  "success": true,
  "total": 1,
}
`;

exports[`sequence webhook should create a user 1`] = `
Object {
  "batch": Array [
    Object {
      "processed": true,
      "success": true,
    },
  ],
  "errors": 0,
  "processed": 1,
  "success": true,
  "total": 1,
}
`;
